{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c51e399c-4e9d-4fc5-ac9d-40e3668051bb",
   "metadata": {},
   "source": [
    "# MMseq2 - preparation / matrix making / exploration of raw output files for mmseq2\n",
    "### most of functions are moved to .py scripts to avoid kernel death\n",
    "* also have analysis of rare biomes\n",
    "* also have analysis of any samples with a poor CDS count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "328d9e73-5a0f-4cfe-bf6f-e477f0cac7e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import datetime\n",
    "import seaborn as sns\n",
    "from sklearn import decomposition\n",
    "import plotly\n",
    "import altair as alt\n",
    "from matplotlib.patches import Patch"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8a0d4a3-58f2-4420-8755-04e61fd14147",
   "metadata": {},
   "source": [
    "# mmseq2 output Data processing"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2756cc74-9703-448a-80a2-82f0f3131abf",
   "metadata": {},
   "source": [
    "input file with the gene and assigned cluster ; python file pulls the sensor from gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "8c97c092-5313-40ef-9e43-184216a37f89",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>bar</th>\n",
       "      <th>gene</th>\n",
       "      <th>is_cluster</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>3</td>\n",
       "      <td>75512.0</td>\n",
       "      <td>Ga0453188_0000018_9739_11091</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>4</td>\n",
       "      <td>75512.0</td>\n",
       "      <td>Ga0453188_0000018_9739_11091</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>5</td>\n",
       "      <td>65133.0</td>\n",
       "      <td>Ga0453188_0000024_43626_45344</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>6</td>\n",
       "      <td>65133.0</td>\n",
       "      <td>Ga0453188_0000024_43626_45344</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7</td>\n",
       "      <td>85135.0</td>\n",
       "      <td>Ga0453188_0000024_43626_45344</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0      bar                           gene  is_cluster\n",
       "0           3  75512.0   Ga0453188_0000018_9739_11091        True\n",
       "1           4  75512.0   Ga0453188_0000018_9739_11091        True\n",
       "2           5  65133.0  Ga0453188_0000024_43626_45344        True\n",
       "3           6  65133.0  Ga0453188_0000024_43626_45344        True\n",
       "4           7  85135.0  Ga0453188_0000024_43626_45344        True"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.read_csv('/global/cfs/cdirs/kbase/KE-Catboost/HK/mode_0/mode_0_clustered_100_for_df.csv',nrows=5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76d4aae0-7184-4e56-9a8a-d5e315178195",
   "metadata": {},
   "outputs": [],
   "source": [
    "x['is_cluster'] = \"True\"\n",
    "x.columns = ['bar', 'gene', 'is_cluster']\n",
    "x.to_csv(\"/global/cfs/cdirs/kbase/KE-Catboost/HK/mode_0/mode_0_clustered_11k_for_df.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d758782-2017-4b99-a093-eb0a6d3b4d73",
   "metadata": {},
   "source": [
    "## 1) Use 'make_matrix.py files in order to pull clusters into catboost matrix\n",
    "* ~500 files will be processed in 3 hours, so broke into ~10 .py scrips to speed the process"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "31c878dd-2465-4f7d-b225-effcd353f3f7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>GOLD Analysis Project ID</th>\n",
       "      <th>Ecosystem</th>\n",
       "      <th>Ecosystem Category</th>\n",
       "      <th>Ecosystem Subtype</th>\n",
       "      <th>Ecosystem Type</th>\n",
       "      <th>Specific Ecosystem</th>\n",
       "      <th>index</th>\n",
       "      <th>406.0</th>\n",
       "      <th>886.0</th>\n",
       "      <th>...</th>\n",
       "      <th>89652.0</th>\n",
       "      <th>94600.0</th>\n",
       "      <th>55402.0</th>\n",
       "      <th>58830.0</th>\n",
       "      <th>75303.0</th>\n",
       "      <th>107579.0</th>\n",
       "      <th>1094.0</th>\n",
       "      <th>94252.0</th>\n",
       "      <th>4964.0</th>\n",
       "      <th>92308.0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>Ga0334848</td>\n",
       "      <td>Environmental</td>\n",
       "      <td>Terrestrial</td>\n",
       "      <td>Wetlands</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Peat</td>\n",
       "      <td>gene_copy</td>\n",
       "      <td>7.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>Ga0334848</td>\n",
       "      <td>Environmental</td>\n",
       "      <td>Terrestrial</td>\n",
       "      <td>Wetlands</td>\n",
       "      <td>Soil</td>\n",
       "      <td>Peat</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>Ga0334884</td>\n",
       "      <td>Engineered</td>\n",
       "      <td>Bioreactor</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Anaerobic</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>gene_copy</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>Ga0334884</td>\n",
       "      <td>Engineered</td>\n",
       "      <td>Bioreactor</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>Anaerobic</td>\n",
       "      <td>Unclassified</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>Ga0325404</td>\n",
       "      <td>Host-associated</td>\n",
       "      <td>Plants</td>\n",
       "      <td>Stem</td>\n",
       "      <td>Phyllosphere</td>\n",
       "      <td>Wood/Secondary xylem</td>\n",
       "      <td>gene_copy</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 14258 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0 GOLD Analysis Project ID        Ecosystem Ecosystem Category  \\\n",
       "0           0                Ga0334848    Environmental        Terrestrial   \n",
       "1           1                Ga0334848    Environmental        Terrestrial   \n",
       "2           0                Ga0334884       Engineered         Bioreactor   \n",
       "3           1                Ga0334884       Engineered         Bioreactor   \n",
       "4           0                Ga0325404  Host-associated             Plants   \n",
       "\n",
       "  Ecosystem Subtype Ecosystem Type    Specific Ecosystem      index  406.0  \\\n",
       "0          Wetlands           Soil                  Peat  gene_copy    7.0   \n",
       "1          Wetlands           Soil                  Peat          0    0.0   \n",
       "2      Unclassified      Anaerobic          Unclassified  gene_copy    5.0   \n",
       "3      Unclassified      Anaerobic          Unclassified          0    0.0   \n",
       "4              Stem   Phyllosphere  Wood/Secondary xylem  gene_copy    0.0   \n",
       "\n",
       "   886.0  ...  89652.0  94600.0  55402.0  58830.0  75303.0  107579.0  1094.0  \\\n",
       "0   20.0  ...      0.0      0.0      0.0      0.0      0.0       0.0     0.0   \n",
       "1    0.0  ...      0.0      0.0      0.0      0.0      0.0       0.0     0.0   \n",
       "2    0.0  ...      0.0      0.0      0.0      0.0      0.0       0.0     0.0   \n",
       "3    0.0  ...      0.0      0.0      0.0      0.0      0.0       0.0     0.0   \n",
       "4    0.0  ...      0.0      0.0      0.0      0.0      0.0       0.0     0.0   \n",
       "\n",
       "   94252.0  4964.0  92308.0  \n",
       "0      0.0     0.0      0.0  \n",
       "1      0.0     0.0      0.0  \n",
       "2      0.0     0.0      0.0  \n",
       "3      0.0     0.0      0.0  \n",
       "4      0.0     0.0      0.0  \n",
       "\n",
       "[5 rows x 14258 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#example output\n",
    "pd.read_csv(\"/global/cfs/cdirs/kbase/KE-Catboost/HK/mode_0_count/copy_matrix_ready_28.csv\",nrows=5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99c8ee63-0593-4567-b1dc-29f782e4c30d",
   "metadata": {},
   "source": [
    "## 2) Use the 'condence_matrix.py' to merge files, add the biome label, remove all 0's\n",
    "## 3) Divide by the total gene_copy_number protein count in all files for abundance"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c04e964-338c-4553-b7e5-d226975f9b16",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "d6441b7f-032b-429d-b79c-2520717640d7",
   "metadata": {},
   "source": [
    "# Mmseq2 data prep\n",
    "* Use the below to read in amino acids in order to do mmseq2\n",
    "* Ended up breaking into different biomes, so that it would not crash the kernel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67b1c6af-3af5-4366-9725-17ef5a8b4379",
   "metadata": {},
   "outputs": [],
   "source": [
    "main_dir = \"/global/cfs/cdirs/kbase/jungbluth/HK_sensors/run2/5000_2/5000_1/\"\n",
    "parent_list = os.listdir(main_dir)\n",
    "my_columns = pd.Series(['GOLD Analysis Project ID', 'gene_oid','pfam_id', 'gene_copy', 'Ecosystem', 'Ecosystem Category',\n",
    "                        'Ecosystem Subtype', 'Ecosystem Type', 'Specific Ecosystem', 'query_start', 'query_end',\n",
    "                       'subj_start', 'subj_end', 'bit_score', 'faa_sequence'], )\n",
    "all_studies_aa = pd.DataFrame()\n",
    "count = 0\n",
    "for child in parent_list:\n",
    "    #if count < 3000: #20,000 total\n",
    "    if 'pfam-hits_table.tsv' in child:\n",
    "            # read in file and extract the biome info\n",
    "        metagenome = pd.read_csv(os.path.join(main_dir,child), delimiter=\"\\t\", usecols=my_columns)\n",
    "        if metagenome.Ecosystem[0] == \"Environmental\":\n",
    "            # remove non-sensory, non-conserved domains\n",
    "            metagenome = metagenome[~metagenome.pfam_id.str.contains(\"pfam02895|pfam18947|pfam00672|pfam01627|pfam07536|pfam00072\")]\n",
    "            # filter to just conserved domains - has to contain both here\n",
    "            just_HK_conserved = metagenome[metagenome.pfam_id.str.contains(\"pfam00512|pfam02518\")]\n",
    "            just_HK_conserved = just_HK_conserved.groupby('gene_oid').count().reset_index()\n",
    "            just_HK_conserved = just_HK_conserved[['gene_oid']].drop_duplicates()\n",
    "            just_HK_conserved[\"has_hk\"] = True\n",
    "            metagenome = metagenome.merge(just_HK_conserved, how='left', on='gene_oid')\n",
    "            metagenome = metagenome[metagenome.has_hk == True]\n",
    "            metagenome = metagenome[~metagenome.pfam_id.str.contains(\"pfam00512|pfam02518\")]\n",
    "            \n",
    "            all_studies_aa = pd.concat([all_studies_aa, metagenome])\n",
    "    #else:\n",
    "       # break\n",
    "    if count%500 == 1:\n",
    "        print(count)\n",
    "        print(datetime.datetime.now())\n",
    "    count = count+1\n",
    "\n",
    "all_studies_aa = all_studies_aa.fillna(0)\n",
    "\n",
    "all_studies_aa.to_csv(\"/global/cfs/cdirs/kbase/KE-Catboost/HK/pfam_amino_acid_all_study_engineered_1.csv\", index='False')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "032d1e13-f21a-43cb-9f56-445c76396a5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(\"/global/cfs/cdirs/kbase/KE-Catboost/HK/matrix_copy_number/catboost_matrix_copy_number.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e124c694-16ef-4c94-b4b1-3d4da6c98311",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_studies_aa = pd.read_csv(\"/global/cfs/cdirs/kbase/KE-Catboost/HK/pfam_amino_acid_all_study_together.csv\")\n",
    "all_studies_aa = all_studies_aa.drop(['Unnamed: 0'], axis=1)\n",
    "all_studies_aa['label_sensory'] = all_studies_aa.pfam_id+\"-\"+all_studies_aa.gene_oid+\"-\"+all_studies_aa.sensory_aa\n",
    "all_studies_aa['biome'] = all_studies_aa['Ecosystem Category']+\":\"+all_studies_aa['Ecosystem Subtype']\n",
    "all_studies_aa.to_csv(\"/global/cfs/cdirs/kbase/KE-Catboost/HK/pfam_amino_acid_all_study_togehter.csv\", index='False')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "91b577d7-f543-4ec5-9b51-4772422ec4cc",
   "metadata": {},
   "source": [
    "#### chop out the sensory domain from each protein"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a3caed7-e4b3-47c0-8b30-264f432e2fb6",
   "metadata": {},
   "outputs": [],
   "source": [
    "sliced = []#pd.DataFrame()\n",
    "i=0\n",
    "for aa in all_studies_aa.faa_sequence:\n",
    "    # cut out query\n",
    "    shortened = aa[all_studies_aa.query_start[i]:all_studies_aa.query_end[i]]\n",
    "    sliced.append(shortened)\n",
    "    i+=1\n",
    "all_studies_aa['sensory_aa'] = sliced"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e497fc1-da94-47b4-8aed-dff6a67b48a4",
   "metadata": {},
   "source": [
    "#### just drop to non-duplicate amino acids to avoid more clustering time. Some gene-pfams have multiple pfam domains/multiple sensory_aa's. So the label must be gene_oid-pfam-sensory_aa to have all genes and all sensory aa in each gene cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fbd87218-7a84-453d-a512-b833baf09d46",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_studies_aa['label_protein'] = all_studies_aa.pfam_id+\"-\"+all_studies_aa.gene_oid\n",
    "all_studies_aa.label_protein.nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a82be1c-4699-47f6-840d-4a051e7c4310",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_studies_aa['biome'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8117efe-66c1-4f71-a102-5e231449a913",
   "metadata": {},
   "source": [
    "Chop out AA corresponding to sensory domain and save as a fasta file"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28d21f3e-299a-4de4-b86b-51dfc2197688",
   "metadata": {},
   "source": [
    "## Make fasta file of sensory domains\n",
    "* make 1 for each biome grouping and then add them all together to avoid crashing the kernel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54e5636f-cdf4-4184-bb0c-c81fa486d4e8",
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta = []\n",
    "i=0\n",
    "for aa in all_studies_aa.label_sensory:\n",
    "    fasta.append(\">%s\\n%s\" % (aa, all_studies_aa.sensory_aa[i]))\n",
    "    i += 1\n",
    "df = pd.Series(fasta).to_csv(index=False)\n",
    "pd.Series(df.replace('\"', '')).to_csv(\"/global/cfs/cdirs/kbase/KE-Catboost/HK/mmseqs/sensory_domains_eng.fasta\", index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4bb9e881-8632-4d67-a2ac-98902c74f9c2",
   "metadata": {},
   "source": [
    "# Processing the cluster file\n",
    "## 1) Read in cluster file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "535e6452-83ae-4435-9a9c-97e7b9120980",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster_base</th>\n",
       "      <th>label_sensory</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pfam13188-CVPL005L_100016672-RIRNATIIKMFDQMSQA...</td>\n",
       "      <td>pfam13188-CVPL005L_100016672-RIRNATIIKMFDQMSQA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pfam13188-CVPL005L_100016672-RIRNATIIKMFDQMSQA...</td>\n",
       "      <td>pfam13188-CVPL005L_100011619-IQMLNQIEQAIVITDQE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>pfam13188-CVPL005L_100016672-RIRNATIIKMFDQMSQA...</td>\n",
       "      <td>pfam13188-CVPL005W_10005012-RIRNATIIKMFDQMSQAI...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pfam13188-CVPL005L_100016672-RIRNATIIKMFDQMSQA...</td>\n",
       "      <td>pfam13188-CVPL010W_100011399-IQMLNQIEQAIVITDQE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>pfam13188-CVPL005L_100016672-RIRNATIIKMFDQMSQA...</td>\n",
       "      <td>pfam13188-CVPL010W_100046382-RIRNATIIKMFDQMSQA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        cluster_base  \\\n",
       "0  pfam13188-CVPL005L_100016672-RIRNATIIKMFDQMSQA...   \n",
       "1  pfam13188-CVPL005L_100016672-RIRNATIIKMFDQMSQA...   \n",
       "2  pfam13188-CVPL005L_100016672-RIRNATIIKMFDQMSQA...   \n",
       "3  pfam13188-CVPL005L_100016672-RIRNATIIKMFDQMSQA...   \n",
       "4  pfam13188-CVPL005L_100016672-RIRNATIIKMFDQMSQA...   \n",
       "\n",
       "                                       label_sensory  \n",
       "0  pfam13188-CVPL005L_100016672-RIRNATIIKMFDQMSQA...  \n",
       "1  pfam13188-CVPL005L_100011619-IQMLNQIEQAIVITDQE...  \n",
       "2  pfam13188-CVPL005W_10005012-RIRNATIIKMFDQMSQAI...  \n",
       "3  pfam13188-CVPL010W_100011399-IQMLNQIEQAIVITDQE...  \n",
       "4  pfam13188-CVPL010W_100046382-RIRNATIIKMFDQMSQA...  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clustered = pd.read_csv('/global/cfs/cdirs/kbase/KE-Catboost/HK/mmseqs/sensory_all_mode0_clu.tsv',\n",
    "                        delimiter='\\t', header=None, names=['cluster_base','label_sensory'])\n",
    "clustered.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e10e1fa9-3620-4c2a-a50f-c5992952706b",
   "metadata": {},
   "source": [
    "# pull out the 'bar' which is the cluster # for each sensory label"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a6fd5b2f-0532-427f-8c43-27ee2375055d",
   "metadata": {},
   "outputs": [],
   "source": [
    "cluster_index = clustered['cluster_base'].drop_duplicates()\n",
    "cluster_index = cluster_index.reset_index(drop=True).rename_axis('bar').reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "36c6d2d1-abfc-4ca0-8002-ffbf100331f0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cluster_base</th>\n",
       "      <th>label_sensory</th>\n",
       "      <th>bar</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pfam13188-CVPL005L_100016672-RIRNATIIKMFDQMSQA...</td>\n",
       "      <td>pfam13188-CVPL005L_100016672-RIRNATIIKMFDQMSQA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pfam13188-CVPL005L_100016672-RIRNATIIKMFDQMSQA...</td>\n",
       "      <td>pfam13188-CVPL005L_100011619-IQMLNQIEQAIVITDQE...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>pfam13188-CVPL005L_100016672-RIRNATIIKMFDQMSQA...</td>\n",
       "      <td>pfam13188-CVPL005W_10005012-RIRNATIIKMFDQMSQAI...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>pfam13188-CVPL005L_100016672-RIRNATIIKMFDQMSQA...</td>\n",
       "      <td>pfam13188-CVPL010W_100011399-IQMLNQIEQAIVITDQE...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>pfam13188-CVPL005L_100016672-RIRNATIIKMFDQMSQA...</td>\n",
       "      <td>pfam13188-CVPL010W_100046382-RIRNATIIKMFDQMSQA...</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        cluster_base  \\\n",
       "0  pfam13188-CVPL005L_100016672-RIRNATIIKMFDQMSQA...   \n",
       "1  pfam13188-CVPL005L_100016672-RIRNATIIKMFDQMSQA...   \n",
       "2  pfam13188-CVPL005L_100016672-RIRNATIIKMFDQMSQA...   \n",
       "3  pfam13188-CVPL005L_100016672-RIRNATIIKMFDQMSQA...   \n",
       "4  pfam13188-CVPL005L_100016672-RIRNATIIKMFDQMSQA...   \n",
       "\n",
       "                                       label_sensory  bar  \n",
       "0  pfam13188-CVPL005L_100016672-RIRNATIIKMFDQMSQA...    0  \n",
       "1  pfam13188-CVPL005L_100011619-IQMLNQIEQAIVITDQE...    0  \n",
       "2  pfam13188-CVPL005W_10005012-RIRNATIIKMFDQMSQAI...    0  \n",
       "3  pfam13188-CVPL010W_100011399-IQMLNQIEQAIVITDQE...    0  \n",
       "4  pfam13188-CVPL010W_100046382-RIRNATIIKMFDQMSQA...    0  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clustered = clustered.merge(cluster_index, on='cluster_base', how='left')\n",
    "clustered.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ecd1bab5-ad69-4157-bcc0-d22b1c7327f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "clustered['pfam_predicted'] = clustered['cluster_base'].str.split('-').str[0]\n",
    "clustered['pfam'] = clustered['label_sensory'].str.split('-').str[0]\n",
    "clustered['gene_predicted'] = clustered['cluster_base'].str.split('-').str[1]\n",
    "clustered['gene_oid'] = clustered['label_sensory'].str.split('-').str[1]\n",
    "clustered = clustered.drop(['cluster_base', 'label_sensory'],axis=1)\n",
    "clustered.to_csv(\"/global/cfs/cdirs/kbase/KE-Catboost/HK/mode_0/clustered_mode_0_saved.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a7378fa-7507-48b3-81ee-d89a5e9f9ce9",
   "metadata": {},
   "source": [
    "# Catboost matrix prep only: \n",
    "* Select only clusters > 100 metagenomes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "852371eb-81c5-4b1f-acd9-f6c1b249918a",
   "metadata": {},
   "source": [
    "# For all matrix:\n",
    "* calculate abundance using make_abundance.py script"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ed0e485-1254-4445-9363-bc323468ca94",
   "metadata": {},
   "outputs": [],
   "source": [
    "groupby_df = all_studies_aa_2.groupby('bar').filter(lambda x : len(x)>100)\n",
    "groupby_df.to_csv(\"/global/cfs/cdirs/kbase/KE-Catboost/HK/clustered_all_over100_metagenomes.csv\")\n",
    "groupby_df = pd.read_csv('/global/cfs/cdirs/kbase/KE-Catboost/HK/clustered_all_over100_metagenomes.csv')\n",
    "groupby_df = groupby_df.drop(['Unnamed: 0'], axis=1)\n",
    "groupby_df.head()\n",
    "\n",
    "groupby_df['is_cluster'] = True\n",
    "groupby_df[['gene', 'bar', 'is_cluster']].to_csv('/global/cfs/cdirs/kbase/KE-Catboost/HK/clustered_100_for_df.csv')\n",
    "\n",
    "groupby_df['is_cluster'] = True\n",
    "groupby_df[['gene', 'bar', 'is_cluster']].to_csv('/global/cfs/cdirs/kbase/KE-Catboost/HK/clustered_100_for_df.csv')\n",
    "groupby_df[['gene', 'bar', 'is_cluster']].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3a97363-180d-4fd0-9d7d-c38471a4e67e",
   "metadata": {},
   "source": [
    "# Visualize the clusters graphically\n",
    "* bring in the biome, and gold ID for all\n",
    "* check for the mmseq2 the % pfam identity in the cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "014b2021-8b0d-4206-8b1c-dac7a4db7cdb",
   "metadata": {},
   "outputs": [],
   "source": [
    "clustered = pd.read_csv(\"/global/cfs/cdirs/kbase/KE-Catboost/HK/mode_0/clustered_mode_0_saved.csv\")\n",
    "clustered = clustered.drop('Unnamed: 0',axis=1).drop_duplicates()\n",
    "cds = pd.read_csv(\"/global/cfs/cdirs/kbase/KE-Catboost/HK/IMG_dataset_Sean_Filtered.csv\")\n",
    "aa = pd.read_csv(\"/global/cfs/cdirs/kbase/KE-Catboost/HK/pfam_amino_acid_all_study_together.csv\",\n",
    "                 usecols = ['gene_oid','Ecosystem','Ecosystem Category','Ecosystem Subtype','GOLD Analysis Project ID'])\n",
    "aa['biome'] = aa['Ecosystem']+\":\"+aa['Ecosystem Category']+\":\"+aa['Ecosystem Subtype']\n",
    "aa = aa[['gene_oid','GOLD Analysis Project ID', 'biome']].drop_duplicates()\n",
    "clustered = clustered.merge(aa, on='gene_oid', how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0a12b3b-aaec-40bc-ae3f-48a1a46c2ff3",
   "metadata": {},
   "outputs": [],
   "source": [
    "print('length:', len(clustered))\n",
    "clustered.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3370b5b8-4908-4a8c-97a6-6a4d543d6fb0",
   "metadata": {},
   "outputs": [],
   "source": [
    "clustered.to_csv(\"/global/cfs/cdirs/kbase/KE-Catboost/HK/mode_0/mode_0_visualized_clusteres.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "de5e41ed-ecf0-4183-b192-9120c53c3c69",
   "metadata": {},
   "source": [
    "### Count percent of pfam that are in the 'right' clusterclustered.pfam_predicted == clustered.pfam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9961eec4-a582-4db6-a710-766106e24fb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "count_pfam_correct = (clustered.pfam_predicted == clustered.pfam).sum()\n",
    "print('pfam correct: ',count_pfam_correct / len(clustered))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3be51497-55d9-49ae-9bfc-97ec3294bd85",
   "metadata": {},
   "source": [
    "# Count of the sensory domains and proteins / biome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "81839518-77b0-4029-8bce-82543951747a",
   "metadata": {},
   "outputs": [],
   "source": [
    "clustered['GOLD Analysis Project ID'].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f79df3ee-817f-4757-88a7-409be9d6b2e9",
   "metadata": {},
   "outputs": [],
   "source": [
    "aa_visualized = groupby_df[['sd_aa', 'biome']]\n",
    "aa_visualized_2 = aa_visualized.drop_duplicates()\n",
    "aa_visualized = aa_visualized.groupby(['sd_aa'])['biome'].nunique()\n",
    "aa_visualized = pd.DataFrame(aa_visualized).reset_index()\n",
    "aa_visualized = aa_visualized.groupby('biome')['sd_aa'].count().reset_index()\n",
    "aa_visualized_2 = groupby_df.groupby(['sd_aa_predicted'])['biome'].nunique()\n",
    "aa_visualized_2 = pd.DataFrame(aa_visualized_2).reset_index()\n",
    "aa_visualized_2 = aa_visualized_2.groupby('biome')['sd_aa_predicted'].count().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "814d8332-a588-47ac-96c4-bffe756ddb1e",
   "metadata": {},
   "outputs": [],
   "source": [
    "aa_visualized['log(unique sensory domains)'] = np.log2(aa_visualized.sd_aa)\n",
    "aa_visualized['log(cluster sensory domains)'] = np.log2(aa_visualized_2.sd_aa_predicted)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "314c8917-c2b3-407e-bfa2-d72507eac30e",
   "metadata": {},
   "outputs": [],
   "source": [
    "aa = aa_visualized.merge(aa_visualized_2, on='biome', how='left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "04bd2ac6-8950-43e8-ab9a-b881d6f76dca",
   "metadata": {},
   "outputs": [],
   "source": [
    "dfm = aa_visualized.drop('sd_aa',axis=1).melt('biome', var_name='cols', value_name='vals')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6b635d86-b506-449b-afe7-8303ab747b74",
   "metadata": {},
   "outputs": [],
   "source": [
    "alt.Chart(dfm, title=\"Count # Biomes Sensory Domain Found In\").mark_circle().encode(\n",
    "    x='biome',\n",
    "    y='vals', tooltip=['biome'],\n",
    "    color='cols').configure_range().properties(width=400,height=300)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4eae2451-1a54-44cc-8f62-9793f0dc440d",
   "metadata": {},
   "source": [
    "# Count the number of unique biomes in each cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "fa608ec6-e88f-47b1-a98d-bf14bed54cad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bar</th>\n",
       "      <th>biome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>4</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>36</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>37</td>\n",
       "      <td>28</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>52</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>65</td>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13497</th>\n",
       "      <td>113110</td>\n",
       "      <td>80</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13498</th>\n",
       "      <td>113127</td>\n",
       "      <td>38</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13499</th>\n",
       "      <td>113135</td>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13500</th>\n",
       "      <td>113166</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13501</th>\n",
       "      <td>113175</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>13502 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          bar  biome\n",
       "0           4     24\n",
       "1          36     48\n",
       "2          37     28\n",
       "3          52     29\n",
       "4          65     24\n",
       "...       ...    ...\n",
       "13497  113110     80\n",
       "13498  113127     38\n",
       "13499  113135     37\n",
       "13500  113166     48\n",
       "13501  113175     10\n",
       "\n",
       "[13502 rows x 2 columns]"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clustered_grouped = clustered.groupby('bar').filter(lambda x : len(x)>100)\n",
    "clustered_grouped = clustered_grouped.groupby(['bar'])['biome'].nunique()\n",
    "clustered_grouped = pd.DataFrame(clustered_grouped).reset_index()\n",
    "clustered_grouped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e15fd79-21f9-4ff9-a4a1-8e91f968cdfb",
   "metadata": {},
   "outputs": [],
   "source": [
    "clustered_grouped = clustered_grouped.groupby('biome')['bar'].count().reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59b87bce-72ed-4811-ab6e-23f8c83e71ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "alt.Chart(clustered_grouped, title=\"Count of Unique Biomes / Cluster\").mark_circle().encode(\n",
    "    x='biome',\n",
    "    y='bar', tooltip=['biome', 'bar'],\n",
    "    color='bar').configure_range(\n",
    "    category={'scheme':'dark2'}\n",
    ").properties(width=400,height=300)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "107958f2-16c2-4f77-82d6-2ddafe25184d",
   "metadata": {},
   "source": [
    "# Rare clusters only in a few biomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "4451298e-5e93-438f-acc5-eb249bf00076",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_63840/1952406927.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  one_two_biomes[\"one_two\"] = True\n",
      "/tmp/ipykernel_63840/1952406927.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  one_two_biomes['biome_count'] = one_two_biomes['biome']\n"
     ]
    }
   ],
   "source": [
    "one_two_biomes = clustered_grouped[clustered_grouped.biome==1]\n",
    "one_two_biomes[\"one_two\"] = True\n",
    "one_two_biomes['biome_count'] = one_two_biomes['biome']\n",
    "cl = clustered.groupby('bar').filter(lambda x : len(x)>100)\n",
    "cl = cl[['bar', 'biome']].drop_duplicates()\n",
    "groupby_df_one_two = cl.merge(one_two_biomes[['bar','one_two', 'biome_count']], how='left', on='bar')\n",
    "\n",
    "groupby_df_one_two = groupby_df_one_two[groupby_df_one_two.one_two == True]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38ce9b51-834e-4f44-a13a-2777602f8c06",
   "metadata": {},
   "source": [
    "### Which clusters < 3 biomes, what are those biomes?\n",
    "#### They are not super spread out - only a few biomesgroupby_df_one_two"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "63736470-93df-452a-ac19-322abf7977ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "groupby_df_one_two"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "a859bb1f-0d2a-44ed-a617-393698b801c8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-d678dcbfc89d40f394d83b7c07bdb889\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  var VEGA_DEBUG = (typeof VEGA_DEBUG == \"undefined\") ? {} : VEGA_DEBUG;\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-d678dcbfc89d40f394d83b7c07bdb889\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-d678dcbfc89d40f394d83b7c07bdb889\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm//vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm//vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm//vega-lite@4.17.0?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm//vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function maybeLoadScript(lib, version) {\n",
       "      var key = `${lib.replace(\"-\", \"\")}_version`;\n",
       "      return (VEGA_DEBUG[key] == version) ?\n",
       "        Promise.resolve(paths[lib]) :\n",
       "        new Promise(function(resolve, reject) {\n",
       "          var s = document.createElement('script');\n",
       "          document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "          s.async = true;\n",
       "          s.onload = () => {\n",
       "            VEGA_DEBUG[key] = version;\n",
       "            return resolve(paths[lib]);\n",
       "          };\n",
       "          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "          s.src = paths[lib];\n",
       "        });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else {\n",
       "      maybeLoadScript(\"vega\", \"5\")\n",
       "        .then(() => maybeLoadScript(\"vega-lite\", \"4.17.0\"))\n",
       "        .then(() => maybeLoadScript(\"vega-embed\", \"6\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 400, \"continuousHeight\": 300}, \"range\": {\"category\": {\"scheme\": \"tableau20\"}}}, \"data\": {\"name\": \"data-67ae6d7f7141b7a18bea6d95ea91a9c5\"}, \"mark\": {\"type\": \"circle\", \"size\": 100}, \"encoding\": {\"color\": {\"field\": \"biome\", \"type\": \"nominal\"}, \"tooltip\": [{\"field\": \"biome\", \"type\": \"nominal\"}, {\"field\": \"bar\", \"type\": \"quantitative\"}], \"x\": {\"field\": \"bar\", \"type\": \"quantitative\"}, \"y\": {\"field\": \"biome\", \"type\": \"nominal\"}}, \"title\": \"Clusters in < 3 Biomes\", \"$schema\": \"https://vega.github.io/schema/vega-lite/v4.17.0.json\", \"datasets\": {\"data-67ae6d7f7141b7a18bea6d95ea91a9c5\": [{\"bar\": 2593, \"biome\": \"Host-associated:Mammals: Human:Large intestine\", \"one_two\": true, \"biome_count\": 1.0}, {\"bar\": 3977, \"biome\": \"Host-associated:Arthropoda: Insects:Gut\", \"one_two\": true, \"biome_count\": 2.0}, {\"bar\": 3977, \"biome\": \"Host-associated:Arthropoda: Insects:Hindgut\", \"one_two\": true, \"biome_count\": 2.0}, {\"bar\": 13670, \"biome\": \"Host-associated:Mammals: Human:Large intestine\", \"one_two\": true, \"biome_count\": 2.0}, {\"bar\": 13670, \"biome\": \"Host-associated:Mammals:Large intestine\", \"one_two\": true, \"biome_count\": 2.0}, {\"bar\": 15032, \"biome\": \"Host-associated:Mammals: Human:Large intestine\", \"one_two\": true, \"biome_count\": 1.0}, {\"bar\": 19260, \"biome\": \"Environmental:Terrestrial:Wetlands\", \"one_two\": true, \"biome_count\": 2.0}, {\"bar\": 19260, \"biome\": \"Host-associated:Mammals: Human:Large intestine\", \"one_two\": true, \"biome_count\": 2.0}, {\"bar\": 19515, \"biome\": \"Environmental:Aquatic:Saline\", \"one_two\": true, \"biome_count\": 2.0}, {\"bar\": 19515, \"biome\": \"Environmental:Aquatic:Hypersaline\", \"one_two\": true, \"biome_count\": 2.0}, {\"bar\": 23349, \"biome\": \"Host-associated:Mammals:Large intestine\", \"one_two\": true, \"biome_count\": 2.0}, {\"bar\": 23349, \"biome\": \"Host-associated:Mammals:Stomach\", \"one_two\": true, \"biome_count\": 2.0}, {\"bar\": 24016, \"biome\": \"Host-associated:Mammals:Stomach\", \"one_two\": true, \"biome_count\": 2.0}, {\"bar\": 24016, \"biome\": \"Host-associated:Mammals:Large intestine\", \"one_two\": true, \"biome_count\": 2.0}, {\"bar\": 30142, \"biome\": \"Host-associated:Mammals:Stomach\", \"one_two\": true, \"biome_count\": 2.0}, {\"bar\": 30142, \"biome\": \"Host-associated:Mammals:Large intestine\", \"one_two\": true, \"biome_count\": 2.0}, {\"bar\": 33030, \"biome\": \"Host-associated:Mammals: Human:Large intestine\", \"one_two\": true, \"biome_count\": 2.0}, {\"bar\": 33030, \"biome\": \"Host-associated:Mammals:Large intestine\", \"one_two\": true, \"biome_count\": 2.0}, {\"bar\": 39242, \"biome\": \"Host-associated:Mammals:Large intestine\", \"one_two\": true, \"biome_count\": 1.0}, {\"bar\": 41140, \"biome\": \"Host-associated:Mammals: Human:Large intestine\", \"one_two\": true, \"biome_count\": 2.0}, {\"bar\": 41140, \"biome\": \"Host-associated:Mammals:Large intestine\", \"one_two\": true, \"biome_count\": 2.0}, {\"bar\": 42217, \"biome\": \"Host-associated:Arthropoda: Insects:Gut\", \"one_two\": true, \"biome_count\": 2.0}, {\"bar\": 42217, \"biome\": \"Host-associated:Arthropoda: Insects:Hindgut\", \"one_two\": true, \"biome_count\": 2.0}, {\"bar\": 43444, \"biome\": \"Host-associated:Mammals: Human:Large intestine\", \"one_two\": true, \"biome_count\": 1.0}, {\"bar\": 43447, \"biome\": \"Host-associated:Mammals: Human:Large intestine\", \"one_two\": true, \"biome_count\": 2.0}, {\"bar\": 43447, \"biome\": \"Environmental:Aquatic:Intertidal zone\", \"one_two\": true, \"biome_count\": 2.0}, {\"bar\": 46798, \"biome\": \"Host-associated:Mammals:Stomach\", \"one_two\": true, \"biome_count\": 2.0}, {\"bar\": 46798, \"biome\": \"Environmental:Terrestrial:Watershed\", \"one_two\": true, \"biome_count\": 2.0}, {\"bar\": 46960, \"biome\": \"Host-associated:Mammals: Human:Large intestine\", \"one_two\": true, \"biome_count\": 1.0}, {\"bar\": 47204, \"biome\": \"Host-associated:Mammals:Large intestine\", \"one_two\": true, \"biome_count\": 2.0}, {\"bar\": 47204, \"biome\": \"Host-associated:Mammals:Stomach\", \"one_two\": true, \"biome_count\": 2.0}, {\"bar\": 47549, \"biome\": \"Environmental:Terrestrial:Insects nest\", \"one_two\": true, \"biome_count\": 1.0}, {\"bar\": 47744, \"biome\": \"Host-associated:Mammals: Human:Large intestine\", \"one_two\": true, \"biome_count\": 2.0}, {\"bar\": 47744, \"biome\": \"Host-associated:Mammals:Large intestine\", \"one_two\": true, \"biome_count\": 2.0}, {\"bar\": 53537, \"biome\": \"Host-associated:Mammals:Stomach\", \"one_two\": true, \"biome_count\": 1.0}, {\"bar\": 55257, \"biome\": \"Host-associated:Mammals: Human:Large intestine\", \"one_two\": true, \"biome_count\": 1.0}, {\"bar\": 61509, \"biome\": \"Host-associated:Mammals: Human:Large intestine\", \"one_two\": true, \"biome_count\": 2.0}, {\"bar\": 61509, \"biome\": \"Engineered:Bioreactor:Unclassified\", \"one_two\": true, \"biome_count\": 2.0}, {\"bar\": 63343, \"biome\": \"Host-associated:Mammals:Stomach\", \"one_two\": true, \"biome_count\": 2.0}, {\"bar\": 63343, \"biome\": \"Host-associated:Mammals:Large intestine\", \"one_two\": true, \"biome_count\": 2.0}, {\"bar\": 63368, \"biome\": \"Host-associated:Mammals: Human:Large intestine\", \"one_two\": true, \"biome_count\": 2.0}, {\"bar\": 63368, \"biome\": \"Host-associated:Mammals:Large intestine\", \"one_two\": true, \"biome_count\": 2.0}, {\"bar\": 64509, \"biome\": \"Host-associated:Mammals: Human:Large intestine\", \"one_two\": true, \"biome_count\": 2.0}, {\"bar\": 64509, \"biome\": \"Engineered:Built environment:Subway\", \"one_two\": true, \"biome_count\": 2.0}, {\"bar\": 74129, \"biome\": \"Host-associated:Mammals: Human:Large intestine\", \"one_two\": true, \"biome_count\": 2.0}, {\"bar\": 74129, \"biome\": \"Host-associated:Mammals:Large intestine\", \"one_two\": true, \"biome_count\": 2.0}, {\"bar\": 76943, \"biome\": \"Host-associated:Mammals: Human:Large intestine\", \"one_two\": true, \"biome_count\": 2.0}, {\"bar\": 76943, \"biome\": \"Host-associated:Mammals:Large intestine\", \"one_two\": true, \"biome_count\": 2.0}, {\"bar\": 77241, \"biome\": \"Host-associated:Mammals: Human:Large intestine\", \"one_two\": true, \"biome_count\": 2.0}, {\"bar\": 77241, \"biome\": \"Host-associated:Mammals:Large intestine\", \"one_two\": true, \"biome_count\": 2.0}, {\"bar\": 77824, \"biome\": \"Host-associated:Mammals:Stomach\", \"one_two\": true, \"biome_count\": 2.0}, {\"bar\": 77824, \"biome\": \"Engineered:WWTP:Unclassified\", \"one_two\": true, \"biome_count\": 2.0}, {\"bar\": 83390, \"biome\": \"Host-associated:Mammals: Human:Large intestine\", \"one_two\": true, \"biome_count\": 1.0}, {\"bar\": 84700, \"biome\": \"Host-associated:Arthropoda: Insects:Gut\", \"one_two\": true, \"biome_count\": 1.0}, {\"bar\": 101165, \"biome\": \"Host-associated:Mammals:Stomach\", \"one_two\": true, \"biome_count\": 2.0}, {\"bar\": 101165, \"biome\": \"Host-associated:Mammals:Large intestine\", \"one_two\": true, \"biome_count\": 2.0}, {\"bar\": 109674, \"biome\": \"Host-associated:Plants:Unclassified\", \"one_two\": true, \"biome_count\": 2.0}, {\"bar\": 109674, \"biome\": \"Environmental:Aquatic:Creek\", \"one_two\": true, \"biome_count\": 2.0}]}}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.Chart(...)"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#groupby_df_one_two['bar'] = np.log2(groupby_df_one_two['bar'])\n",
    "alt.Chart(groupby_df_one_two, title=\"Clusters in < 3 Biomes\").mark_circle(size=100).encode(\n",
    "    x=alt.X('bar'),#scale=alt.Scale(domain=[10, 16])),\n",
    "    y='biome', tooltip=['biome', 'bar'],\n",
    "    color='biome').configure_range(\n",
    "    category={'scheme':'tableau20'}\n",
    ")#.configure_axis(\n",
    "  #  labelFontSize=14,\n",
    "   # titleFontSize=14).interactive()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5771ac43-8814-4318-9fb9-700f308537a2",
   "metadata": {},
   "source": [
    "#### The clusters for algae, insect, bird:large intestine, plant leaf?\n",
    "\n",
    "#### 4352, 19268, 4490, 14283, 13260, 17741, 16247"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "685ed7e5-37c0-4e14-8f51-1bfe89689b43",
   "metadata": {},
   "source": [
    "### Which clusters == 1 biomes, what are those biomes?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "f0dbed4c-e22d-4b2e-b55a-2c16c66fd671",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<div id=\"altair-viz-22e7dc42483d4555aa60dd3803553c9d\"></div>\n",
       "<script type=\"text/javascript\">\n",
       "  var VEGA_DEBUG = (typeof VEGA_DEBUG == \"undefined\") ? {} : VEGA_DEBUG;\n",
       "  (function(spec, embedOpt){\n",
       "    let outputDiv = document.currentScript.previousElementSibling;\n",
       "    if (outputDiv.id !== \"altair-viz-22e7dc42483d4555aa60dd3803553c9d\") {\n",
       "      outputDiv = document.getElementById(\"altair-viz-22e7dc42483d4555aa60dd3803553c9d\");\n",
       "    }\n",
       "    const paths = {\n",
       "      \"vega\": \"https://cdn.jsdelivr.net/npm//vega@5?noext\",\n",
       "      \"vega-lib\": \"https://cdn.jsdelivr.net/npm//vega-lib?noext\",\n",
       "      \"vega-lite\": \"https://cdn.jsdelivr.net/npm//vega-lite@4.17.0?noext\",\n",
       "      \"vega-embed\": \"https://cdn.jsdelivr.net/npm//vega-embed@6?noext\",\n",
       "    };\n",
       "\n",
       "    function maybeLoadScript(lib, version) {\n",
       "      var key = `${lib.replace(\"-\", \"\")}_version`;\n",
       "      return (VEGA_DEBUG[key] == version) ?\n",
       "        Promise.resolve(paths[lib]) :\n",
       "        new Promise(function(resolve, reject) {\n",
       "          var s = document.createElement('script');\n",
       "          document.getElementsByTagName(\"head\")[0].appendChild(s);\n",
       "          s.async = true;\n",
       "          s.onload = () => {\n",
       "            VEGA_DEBUG[key] = version;\n",
       "            return resolve(paths[lib]);\n",
       "          };\n",
       "          s.onerror = () => reject(`Error loading script: ${paths[lib]}`);\n",
       "          s.src = paths[lib];\n",
       "        });\n",
       "    }\n",
       "\n",
       "    function showError(err) {\n",
       "      outputDiv.innerHTML = `<div class=\"error\" style=\"color:red;\">${err}</div>`;\n",
       "      throw err;\n",
       "    }\n",
       "\n",
       "    function displayChart(vegaEmbed) {\n",
       "      vegaEmbed(outputDiv, spec, embedOpt)\n",
       "        .catch(err => showError(`Javascript Error: ${err.message}<br>This usually means there's a typo in your chart specification. See the javascript console for the full traceback.`));\n",
       "    }\n",
       "\n",
       "    if(typeof define === \"function\" && define.amd) {\n",
       "      requirejs.config({paths});\n",
       "      require([\"vega-embed\"], displayChart, err => showError(`Error loading script: ${err.message}`));\n",
       "    } else {\n",
       "      maybeLoadScript(\"vega\", \"5\")\n",
       "        .then(() => maybeLoadScript(\"vega-lite\", \"4.17.0\"))\n",
       "        .then(() => maybeLoadScript(\"vega-embed\", \"6\"))\n",
       "        .catch(showError)\n",
       "        .then(() => displayChart(vegaEmbed));\n",
       "    }\n",
       "  })({\"config\": {\"view\": {\"continuousWidth\": 400, \"continuousHeight\": 300}, \"range\": {\"category\": {\"scheme\": \"tableau20\"}}}, \"data\": {\"name\": \"data-141f1ae875db64cd2758e849e79d24c8\"}, \"mark\": \"circle\", \"encoding\": {\"color\": {\"field\": \"biome\", \"type\": \"nominal\"}, \"tooltip\": [{\"field\": \"biome\", \"type\": \"nominal\"}, {\"field\": \"bar\", \"type\": \"quantitative\"}], \"x\": {\"field\": \"bar\", \"type\": \"quantitative\"}, \"y\": {\"field\": \"biome\", \"type\": \"nominal\"}}, \"title\": \"Clusters in 1 Biomes\", \"$schema\": \"https://vega.github.io/schema/vega-lite/v4.17.0.json\", \"datasets\": {\"data-141f1ae875db64cd2758e849e79d24c8\": [{\"bar\": 2593, \"biome\": \"Host-associated:Mammals: Human:Large intestine\", \"one_two\": true, \"biome_count\": 1.0}, {\"bar\": 15032, \"biome\": \"Host-associated:Mammals: Human:Large intestine\", \"one_two\": true, \"biome_count\": 1.0}, {\"bar\": 39242, \"biome\": \"Host-associated:Mammals:Large intestine\", \"one_two\": true, \"biome_count\": 1.0}, {\"bar\": 43444, \"biome\": \"Host-associated:Mammals: Human:Large intestine\", \"one_two\": true, \"biome_count\": 1.0}, {\"bar\": 46960, \"biome\": \"Host-associated:Mammals: Human:Large intestine\", \"one_two\": true, \"biome_count\": 1.0}, {\"bar\": 47549, \"biome\": \"Environmental:Terrestrial:Insects nest\", \"one_two\": true, \"biome_count\": 1.0}, {\"bar\": 53537, \"biome\": \"Host-associated:Mammals:Stomach\", \"one_two\": true, \"biome_count\": 1.0}, {\"bar\": 55257, \"biome\": \"Host-associated:Mammals: Human:Large intestine\", \"one_two\": true, \"biome_count\": 1.0}, {\"bar\": 83390, \"biome\": \"Host-associated:Mammals: Human:Large intestine\", \"one_two\": true, \"biome_count\": 1.0}, {\"bar\": 84700, \"biome\": \"Host-associated:Arthropoda: Insects:Gut\", \"one_two\": true, \"biome_count\": 1.0}]}}, {\"mode\": \"vega-lite\"});\n",
       "</script>"
      ],
      "text/plain": [
       "alt.Chart(...)"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "alt.Chart(groupby_df_one_two, title=\"Clusters in 1 Biomes\").mark_circle().encode(\n",
    "    x='bar',\n",
    "    y='biome', tooltip=['biome', 'bar'],\n",
    "    color='biome').configure_range(\n",
    "    category={'scheme':'tableau20'}\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1c9151e-60fb-4840-88a0-cc7bd38d4fbb",
   "metadata": {},
   "source": [
    "# Which genomes have a bad CDS count?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ccdc0097-cd74-4140-a5d1-de069b7698ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "cds = pd.read_csv(\"/global/cfs/cdirs/kbase/KE-Catboost/HK/IMG_dataset_Sean_Filtered.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57d87c68-ce99-4f9b-b317-f7dc26ef4d9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "cds['assembled.CDS.Count']\n",
    "\n",
    "bin_labels_5 = ['<25k', '<50k', '<100k', '>100k']\n",
    "cds['CDS_bins'] = pd.cut(cds['assembled.CDS.Count'],\n",
    "                              bins=[-1, 25000, 50000, 100000, np.inf],\n",
    "                              labels=bin_labels_5)\n",
    "cds.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1cbd65c2-a5ed-4bb9-a8ff-12470f066097",
   "metadata": {},
   "outputs": [],
   "source": [
    "cds = cds.groupby('biome').filter(lambda x : len(x) > 30)\n",
    "cds_grouped = pd.DataFrame(cds[['biome','CDS_bins']].groupby(['biome','CDS_bins'])['CDS_bins'].count())\n",
    "cds_grouped = cds_grouped.rename(columns={\"CDS_bins\": \"count\"}).reset_index()\n",
    "cds_grouped = cds_grouped[cds_grouped['count'] > 0].reset_index(drop=True)\n",
    "cds_grouped.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0f4f993-8340-402a-b11d-9586e19231ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "cds_grouped = cds_grouped[cds_grouped.CDS_bins!='>100k']\n",
    "alt.Chart(cds_grouped[cds_grouped['count']>20]).mark_bar().encode(\n",
    "    y='biome',\n",
    "    x='count',\n",
    "    color='CDS_bins').configure_range(\n",
    "    category={'scheme':'tableau20'}\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a247fd4-ad32-4562-8fed-0054ea507106",
   "metadata": {},
   "outputs": [],
   "source": [
    "alt.Chart(cds_grouped[cds_grouped.CDS_bins=='<25k']).mark_arc(innerRadius=50).encode(\n",
    "    theta=alt.Theta(field=\"count\", type=\"quantitative\"),\n",
    "    color=alt.Color(field=\"biome\", type=\"nominal\"),\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "943eafbd-bda6-41f2-810b-4945a3ce67e7",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "NERSC Python",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
